### ISSUE TITLE
panic: runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by dongmingh** [hyperledger-archives/fabric#1447]
## Description

commit level: 0bbda1a20aba577c7eeec8e37b8963c5934ee554
When measuring the TPS for hyperledger fabric with various pbft modes, observed panic twice with pbft classic mode with security disabled.  At the time of observing, the logging level was set to error.  Then I attempted to recreate with logging level set to debug, but I could not see the failure again.  Then I switched the logging level back to error and saw it again (only once) and gathered the back trace below.  Then I do not see it anymore.  Hence this is an inter-mitten and may be difficult to recreate.  I will post information here if I see it again.

The back trace:

panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x8416ff]

goroutine 12 [running]:
panic(0xd30380, 0xc8200100b0)
    /opt/go/src/runtime/panic.go:464 +0x3e6
github.com/hyperledger/fabric/core/ledger.(_blockchain).getBlockchainInfoForBlock(0xc8202210c0, 0x15, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/blockchain.go:179 +0xcf
github.com/hyperledger/fabric/core/ledger.(_blockchain).getBlockchainInfo(0xc8202210c0, 0x84705b, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/blockchain.go:170 +0xdb
github.com/hyperledger/fabric/core/ledger.(_Ledger).GetBlockchainInfo(0xc820258a20, 0x0, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/ledger.go:324 +0x30
github.com/hyperledger/fabric/consensus/helper.(_Helper).GetBlockchainInfoBlob(0xc8200a9f80, 0x0, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/consensus/helper/helper.go:398 +0x3a
github.com/hyperledger/fabric/consensus/obcpbft.(_obcGeneric).getState(0xc820272180, 0x0, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/obc-pbft.go:137 +0x4a
github.com/hyperledger/fabric/consensus/obcpbft.(_pbftCore).execDoneSync(0xc82007b1e0)
    /opt/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:892 +0x1c3
github.com/hyperledger/fabric/consensus/obcpbft.(*pbftCore).main(0xc82007b1e0)
    /opt/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:294 +0x7f3
created by github.com/hyperledger/fabric/consensus/obcpbft.newPbftCore
    /opt/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:254 +0x1632
## Describe How to Reproduce
1. execute command "go test -run=BuildImage_Peer" in core/container
2. Use script, see local_fabric.sh in https://github.com/rameshthoomu/fabric/tree/tools/localpeersetup. 
   Set the consensus to pbft and mode to classic. I changed two lines in the script in order to use the peer created from step 1:

PEER_IMAGE=hyperledger-peer:latest
MEMBERSRVC_IMAGE=membersrvc:latest

The commands to launch 4 peers vagrant containers:
./local_fabric.sh -n 4 (security disabled)
./local_fabric.sh -n 4 -s (security enabled)
1. Use SoapUI to generate transactions with 100 threads.
2. The panic occurs when sending invokes with security disabled.

**Comment by muralisrini** [comment]
@dongmingh : what transactions do we generate in 1. ? We know deploy transactions can cause problems with ledger or image creation if there are large files (such as those logs from behave tests) under fabric folder. For a reference point, I saw issues with 0.5Gb image size.

**Comment by dongmingh** [comment]
After launch 4 peers, I used SoapUI to generate
1. deploy chaincode example02
2. two invokes (a->b and b->a) and two queries (a and b) transactions, one tansaction at a time to confirm all transactions work correctly
3. run invokes (a->b, then b->a repeatedly) with 100 threads for 5 minutes.  The panic occurred in the beginning of the run.

**Comment by corecode** [comment]
@srderson any idea how this can happen?  I don't see what could even cause this.  Why is *block == 0x0?

@dongmingh do you have a full stack trace, not just of the panicking goroutine?

**Comment by dongmingh** [comment]
@corecode I have put all back trace when I opened the bug.

**Comment by srderson** [comment]
Not sure. I took a quick look and it wasn't clear. @manish-sethi any ideas?

@corecode Is consensus calling ledger.PutRawBlock at some point?

**Comment by manish-sethi** [comment]
No clue what could cause this... especially last "0x0"  (after *block address) haven't see for a pointer type return value.
@dongmingh by any chance do you have debug logs for this run?

**Comment by dongmingh** [comment]
I found very limited log, see below, since the logging level was set to error.  I was trying to recreate with logging level set to debug, but had no luck.
peer 0:
17:07:23.948 [crypto] main -> INFO 001[0m Log level recognized 'info', set to INFO
[31m17:08:49.172 [rest] Query -> ERRO 002[0m {"Error": "Querying Chaincode -- Error:Failed to launch chaincode spec(Could not get deployment transaction for 1edd7021ab71b766f4928a9ef91182c018dffb86fef7a4b5a5516ac590a87957e21a62d939df817f5105f524abddcddfc7b1a60d780f02d8235bd7af9db81b66 - ledger: resource not found)"}
[31m17:10:37.343 [protos] Bytes -> ERRO 003[0m Error marshalling block: proto: Marshal called with nil
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x8416ff]

goroutine 12 [running]:
panic(0xd30380, 0xc8200100b0)
    /opt/go/src/runtime/panic.go:464 +0x3e6
github.com/hyperledger/fabric/core/ledger.(_blockchain).getBlockchainInfoForBlock(0xc8202210c0, 0x15, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/blockchain.go:179 +0xcf
github.com/hyperledger/fabric/core/ledger.(_blockchain).getBlockchainInfo(0xc8202210c0, 0x84705b, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/blockchain.go:170 +0xdb
github.com/hyperledger/fabric/core/ledger.(_Ledger).GetBlockchainInfo(0xc820258a20, 0x0, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/ledger.go:324 +0x30
github.com/hyperledger/fabric/consensus/helper.(_Helper).GetBlockchainInfoBlob(0xc8200a9f80, 0x0, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/consensus/helper/helper.go:398 +0x3a
github.com/hyperledger/fabric/consensus/obcpbft.(_obcGeneric).getState(0xc820272180, 0x0, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/obc-pbft.go:137 +0x4a
github.com/hyperledger/fabric/consensus/obcpbft.(_pbftCore).execDoneSync(0xc82007b1e0)
    /opt/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:892 +0x1c3
github.com/hyperledger/fabric/consensus/obcpbft.(*pbftCore).main(0xc82007b1e0)
    /opt/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:294 +0x7f3
created by github.com/hyperledger/fabric/consensus/obcpbft.newPbftCore
    /opt/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:254 +0x1632

peer 1:
17:07:23.947 [crypto] main -> INFO 001[0m Log level recognized 'info', set to INFO
[31m17:07:31.477 [peer] handleChat -> ERRO 002[0m Error handling message: Peer FSM failed while handling message (DISC_HELLO): current state: created, error: transition canceled with error: Error registering Handler: Duplicate Handler error: {name:"vp1"  172.17.0.1:30003 VALIDATOR }
[31m17:07:31.577 [peer] handleChat -> ERRO 003[0m Error handling message: Peer FSM failed while handling message (DISC_HELLO): current state: created, error: transition canceled with error: Error registering Handler: Duplicate Handler error: {name:"vp3"  172.17.0.1:30007 VALIDATOR }
2016/05/09 17:10:37 transport: http2Client.notifyError got notified that the client transport was broken EOF.
[31m17:10:37.462 [peer] handleChat -> ERRO 004[0m Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
[31m17:10:37.463 [peer] chatWithPeer -> ERRO 005[0m Ending chat with peer address=172.17.0.1:30001 due to error:  Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
2016/05/09 17:10:37 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.4:43535->172.17.0.1:30001: read: connection reset by peer.
2016/05/09 17:10:37 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.4:43541->172.17.0.1:30001: read: connection reset by peer.
2016/05/09 17:10:38 grpc: ClientConn.resetTransport failed to create client transport: connection error: desc = "transport: write tcp 172.17.0.4:43543->172.17.0.1:30001: write: broken pipe"; Reconnecting to "172.17.0.1:30001"
2016/05/09 17:10:40 grpc: ClientConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.17.0.1:30001: getsockopt: connection refused"; Reconnecting to "172.17.0.1:30001"
2016/05/09 17:10:40 grpc: ClientConn.transportMonitor exits due to: grpc: timed out trying to connect
[31m17:11:48.489 [chaincode] processStream -> ERRO 006[0m Error handling chaincode support stream: connection error: desc = "transport is closing"
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken EOF.
[31m17:11:48.764 [peer] handleChat -> ERRO 007[0m Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
[31m17:11:48.765 [peer] chatWithPeer -> ERRO 008[0m Ending chat with peer address=172.17.0.1:30007 due to error:  Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
[31m17:11:48.765 [peer] handleChat -> ERRO 009[0m Error during Chat, stopping handler: connection error: desc = "transport is closing"
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.4:33456->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.4:33458->172.17.0.1:30007: read: connection reset by peer.

peer 2:
**17:07:24.352 [crypto] main -> INFO 001[0m Log level recognized 'info', set to INFO
[31m17:07:31.578 [peer] handleChat -> ERRO 002[0m Error handling message: Peer FSM failed while handling message (DISC_HELLO): current state: created, error: transition canceled with error: Error registering Handler: Duplicate Handler error: {name:"vp2"  172.17.0.1:30005 VALIDATOR }
[31m17:07:31.581 [peer] handleChat -> ERRO 003[0m Error handling message: Peer FSM failed while handling message (DISC_HELLO): current state: created, error: transition canceled with error: Error registering Handler: Duplicate Handler error: {name:"vp1"  172.17.0.1:30003 VALIDATOR }
2016/05/09 17:10:37 transport: http2Client.notifyError got notified that the client transport was broken EOF.
[31m17:10:37.486 [peer] handleChat -> ERRO 004[0m Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
[31m17:10:37.487 [peer] chatWithPeer -> ERRO 005[0m Ending chat with peer address=172.17.0.1:30001 due to error:  Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
[31m17:10:37.487 [peer] sendBlocks -> ERRO 006[0m Error sending blockNum 4: Error Sending message through ChatStream: rpc error: code = 13 desc = "transport is closing"
2016/05/09 17:10:37 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.5:41320->172.17.0.1:30001: read: connection reset by peer.
2016/05/09 17:10:37 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.5:41326->172.17.0.1:30001: read: connection reset by peer.
2016/05/09 17:10:38 grpc: ClientConn.resetTransport failed to create client transport: connection error: desc = "transport: write tcp 172.17.0.5:41328->172.17.0.1:30001: write: broken pipe"; Reconnecting to "172.17.0.1:30001"
2016/05/09 17:10:40 grpc: ClientConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.17.0.1:30001: getsockopt: connection refused"; Reconnecting to "172.17.0.1:30001"
2016/05/09 17:10:40 grpc: ClientConn.transportMonitor exits due to: grpc: timed out trying to connect
[31m17:11:48.572 [chaincode] processStream -> ERRO 007[0m Error handling chaincode support stream: connection error: desc = "transport is closing"

peer 3:
17:07:23.948 [crypto] main -> INFO 001[0m Log level recognized 'info', set to INFO
[31m17:07:31.475 [peer] handleChat -> ERRO 002[0m Error handling message: Peer FSM failed while handling message (DISC_HELLO): current state: created, error: transition canceled with error: Error registering Handler: Duplicate Handler error: {name:"vp2"  172.17.0.1:30005 VALIDATOR }
[31m17:07:31.575 [peer] handleChat -> ERRO 003[0m Error handling message: Peer FSM failed while handling message (DISC_HELLO): current state: created, error: transition canceled with error: Error registering Handler: Duplicate Handler error: {name:"vp3"  172.17.0.1:30007 VALIDATOR }
2016/05/09 17:10:37 transport: http2Client.notifyError got notified that the client transport was broken EOF.
[31m17:10:37.459 [peer] handleChat -> ERRO 004[0m Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
[31m17:10:37.464 [peer] chatWithPeer -> ERRO 005[0m Ending chat with peer address=172.17.0.1:30001 due to error:  Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
2016/05/09 17:10:37 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:51441->172.17.0.1:30001: read: connection reset by peer.
2016/05/09 17:10:38 grpc: ClientConn.resetTransport failed to create client transport: connection error: desc = "transport: write tcp 172.17.0.3:51454->172.17.0.1:30001: write: broken pipe"; Reconnecting to "172.17.0.1:30001"
2016/05/09 17:10:40 grpc: ClientConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.17.0.1:30001: getsockopt: connection refused"; Reconnecting to "172.17.0.1:30001"
2016/05/09 17:10:40 grpc: ClientConn.transportMonitor exits due to: grpc: timed out trying to connect
[31m17:11:48.664 [chaincode] processStream -> ERRO 006[0m Error handling chaincode support stream: connection error: desc = "transport is closing"
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken EOF.
[31m17:11:48.765 [peer] handleChat -> ERRO 007[0m Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47755->172.17.0.1:30007: read: connection reset by peer.
[31m17:11:48.766 [peer] chatWithPeer -> ERRO 009[0m Ending chat with peer address=172.17.0.1:30007 due to error:  Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
[31m17:11:48.766 [peer] handleChat -> ERRO 008[0m Error during Chat, stopping handler: connection error: desc = "transport is closing"
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47758->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47762->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47764->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47766->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47768->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47770->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47772->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47774->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47776->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47778->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47780->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47782->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47784->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47786->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47788->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47790->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47792->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47794->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47796->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47798->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47800->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47802->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47804->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47806->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47808->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47810->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47812->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47814->172.17.0.1:30007: read: connection reset by peer.
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.17.0.3:47816->172.17.0.1:30007: read: connection reset by peer.
[31m17:11:48.996 [peer] handleChat -> ERRO 00a[0m Error during Chat, stopping handler: connection error: desc = "transport is closing"
2016/05/09 17:11:48 transport: http2Client.notifyError got notified that the client transport was broken EOF.
[31m17:11:48.998 [peer] handleChat -> ERRO 00b[0m Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"
[31m17:11:48.998 [peer] chatWithPeer -> ERRO 00c[0m Ending chat with peer address=172.17.0.1:30005 due to error:  Error during Chat, stopping handler: rpc error: code = 13 desc = "transport is closing"

**Comment by yacovm** [comment]
in getBlockchainInfo, if the lastBlock is nill it gets passed to getBlockchainInfoForBlock and there a BlockchainInfo is built with block.PreviousBlockHash which might throw the NPE in my opinion.

**Comment by manish-sethi** [comment]
@yacovm if lastBlock is nil that would mean that the blockchain size is zero (i.e., there is no block in chain yet). In that case, the method explicitly returns `&protos.BlockchainInfo{Height: 0}`.

**Comment by yacovm** [comment]
My bad, you're right.

**Comment by christo4ferris** [comment]
pretty sure that this is no longer relevant. @corecode would you please close if you agree. thx

**Comment by dongmingh** [comment]
Have not seen this for a long time.  Close it now.
